src/ml-service/packages/sensor_location/locate/app.py

```diff
@@ -0,0 +1,152 @@
+    '''
+    district = 'WAKISO'
+    subcounty = None
+    sensor_number = None
+
+    all_parishes = locate_model.get_parishes(district, subcounty)
+    if len(all_parishes) < 2:
+        return 'Invalid input data'
+    else:
+        all_parishes_df = locate_helper.json_to_df(all_parishes)
+        all_parishes_df = locate_helper.process_data(all_parishes_df)
+        recommended_parishes = locate_helper.kmeans_algorithm(
+            all_parishes_df, sensor_number)
+        return jsonify(recommended_parishes)
+
+
+@app.route('/api/v1/map/parishes', methods=['POST'])
+def place_sensors_map():
+    '''
+    Returns parishes recommended by the model given the polygon and must-have coordinates
+    '''
+    if request.method == 'POST':
+        json_data = request.get_json()
+        if not json_data:
+            return {'message': 'No input data provided'}, 400
+        else:
+            sensor_number = int(json_data["sensor_number"])
+
+            polygon = json_data["polygon"]
+            if polygon == {}:
+                return {'message': 'Please draw a polygon'}, 200
+            geometry = polygon["geometry"]["coordinates"]
+
+            must_have_coordinates = json_data["must_have_coordinates"]
+            if must_have_coordinates == "":
+                must_have_coordinates = None
+                return locate_helper.recommend_locations(sensor_number, must_have_coordinates, geometry)
+            else:
+                try:
+                    must_have_coordinates = ast.literal_eval(
+                        must_have_coordinates)
+                except:
+                    return {'message': 'Coordinates must be in the form [[long, lat], [long, lat]]'}, 200
+                if all(isinstance(x, list) for x in must_have_coordinates):
+                    return locate_helper.recommend_locations(sensor_number, must_have_coordinates, geometry)
+                else:
+                    return {'message': 'Coordinates must be in the form [[longitude, latitude]]'}, 200
+
+
+@app.route('/api/v1/map/savelocatemap', methods=['GET', 'POST'])
+def save_locate_map():
+    '''
+    Saves planning space
+    '''
+    if request.content_type != 'application/json':
+        error = json.dumps({'error': 'Invalid Content Type'})
+        return jsonify(error, 400)
+
+    data = request.json
+    if not all([data.get('user_id'), data.get('space_name'), data.get('plan')]):
+        error = json.dumps({'error': 'Missing field/s (title, author_id)'})
+        return jsonify(error, 400)
+
+    user_id = data['user_id']
+    space_name = data['space_name']
+    plan = data['plan']
+
+    locate_model.save_locate_map(user_id, space_name, plan)
+
+    return jsonify({"message": "Locate Plannig Space Saved Successfully", "status": 200})
+
+# get previously saved planning space by the current user
+@app.route('/api/v1/map/getlocatemap/<user_id>')
+def get_locate_map(user_id):
```

Comment:
Perhaps be able to check for the existence of the userID in the table? We could make call to the Users document inside airqo_analytics. Since we are dealing with different databases now, we could make use of an API endpoint for finding a user - this could the one called: Get One User (/api/v1/users/<user ID>), more details can be doing in the auth-service api doc sheet.

On reviewing the ATLAS collections, I can see three: **locate_map**, **geometry_polygon** and **geo_census**, which of these 3 are we using to store the planning spaces?

Reply:
Thanks @Baalmart for the awesome reviews. I have taken note of them. For the collection, planning spaces are stored in *locate_map*.

Reply:
> For some reason, the response is always an empty array whenever I save a planning space.

Hi @Baalmart, this has been resolved in the new push